### DISCUSSION TITLE
File "microdot.py", line 25 SyntaxError: invalid syntax

### DISCUSSION THREAD

**Discussion by QGB** [miguelgrinberg/microdot#110] [Q&A]
```
>>> import microdot
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "microdot.py", line 25
SyntaxError: invalid syntax

```
upip install to /Lib/src ,so I directly save https://github.com/miguelgrinberg/microdot/blob/main/src/microdot.py to esp32c3 board

**Comment by miguelgrinberg** [comment]
What do you have in line 25 of microdot.py? Do not look in github please, look in the copy of the file that was installed on your device.

**Comment by QGB** [comment]
@miguelgrinberg 
```
import upip
upip.install('microdot')
```

it is the correct way to install microdot in esp32 board?

**Comment by miguelgrinberg** [comment]
@QGB This is what I have in the documentation wrt installation:

> For MicroPython, you can install it with upip if that option is available, but the recommended approach is to manually copy microdot.py and any desired optional extension source files from the [GitHub repository](https://github.com/miguelgrinberg/microdot/tree/main/src) into your device, possibly after [compiling](https://docs.micropython.org/en/latest/reference/mpyfiles.html) them to .mpy files. These source files can also be [frozen](https://docs.micropython.org/en/latest/develop/optimizations.html?highlight=frozen#frozen-bytecode) and incorporated into a custom MicroPython firmware.

I don't use `upip`, and besides, it is getting replaced with a different tool called `mip`. I suggest you copy the files that you need manually, if possible after compiling them with mpy-cross so that they are smaller.
